crates/symbolicator-service/src/services/download/mod.rs

```diff
@@ -82,6 +83,114 @@ impl From<GcsError> for CacheError {
     }
 }
 
+/// A record of a number of download failures in a given second.
+#[derive(Debug, Clone, Copy)]
+struct FailureCount {
+    /// The time at which the failures occurred, measured since the Unix Epoch.
+    timestamp: Duration,
+    /// The number of failures.
+    failures: usize,
+}
+
+type CountedFailures = Arc<Mutex<VecDeque<FailureCount>>>;
+
+/// A structure that keeps track of download failures in a given time interval
+/// and puts hosts on a block list accordingly.
+///
+/// The logic works like this: if a host has at least `FAILURE_THRESHOLD` download
+/// failures in a window of `TIME_WINDOW` seconds, it will be blocked for a duration of
+/// `BLOCK_TIME`.
+#[derive(Clone, Debug)]
+struct HostDenyList {
+    time_window: Duration,
```

Comment:
Not quite sure what `Duration` is internally, but it might be worth just using `u64` millis internally.
We would then only need a single `as_millis` per `round_duration`, as all the others happen in the constructor.

Reply:
Alright, I changed it so that internally we only use `u64` milliseconds.